server/src/main/java/org/elasticsearch/action/ingest/SimulatePipelineResponse.java

```diff
@@ -19,22 +19,87 @@
+import org.elasticsearch.common.xcontent.ConstructingObjectParser;
 import org.elasticsearch.common.xcontent.ToXContentObject;
 import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.XContentParser.Token;
 
 import java.io.IOException;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
 
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.constructorArg;
+import static org.elasticsearch.common.xcontent.XContentParserUtils.ensureExpectedToken;
+
 public class SimulatePipelineResponse extends ActionResponse implements ToXContentObject {
     private String pipelineId;
     private boolean verbose;
     private List<SimulateDocumentResult> results;
 
+    @SuppressWarnings("unchecked")
+    public static final ConstructingObjectParser<SimulatePipelineResponse, Void> PARSER =
+        new ConstructingObjectParser<>(
+            "simulate_pipeline_response",
+            true,
+            a -> {
+                List<SimulateDocumentResult> results = (List<SimulateDocumentResult>)a[0];
+                boolean verbose = false;
+                if (results.size() > 0) {
+                    if (results.get(0) instanceof SimulateDocumentVerboseResult) {
+                        verbose = true;
+                    }
+                }
+                return new SimulatePipelineResponse(null, verbose, results);
+            }
+        );
+    static {
+        PARSER.declareObjectArray(
+            constructorArg(),
+            (parser, context) -> {
+                ensureExpectedToken(Token.START_OBJECT, parser.currentToken(), parser::getTokenLocation);
+                SimulateDocumentResult result = null;
+                while (parser.nextToken() != Token.END_OBJECT) {
+                    ensureExpectedToken(parser.currentToken(), Token.FIELD_NAME, parser::getTokenLocation);
+                    String fieldName = parser.currentName();
+                    parser.nextToken();
+                    switch(fieldName) {
+                        case SimulateDocumentVerboseResult.PROCESSOR_RESULT_FIELD:
+                            ensureExpectedToken(Token.START_ARRAY, parser.currentToken(), parser::getTokenLocation);
+                            List<SimulateProcessorResult> results = new ArrayList<>();
+                            while (parser.nextToken().equals(Token.START_OBJECT)) {
+                                results.add(SimulateProcessorResult.fromXContent(parser));
+                            }
+                            ensureExpectedToken(Token.END_ARRAY, parser.currentToken(), parser::getTokenLocation);
+                            result = new SimulateDocumentVerboseResult(results);
+                            break;
+                        case WriteableIngestDocument.DOC_FIELD:
+                        case "error":
```

Comment:
maybe have two cases for this rather than a single case with an if that checks again?

Reply:
Yes, this made more sense before when it was a bit different. I will change this.